app/vmagent/remotewrite/client.go

```diff
@@ -462,21 +462,25 @@ again:
 		// - Real-world implementations of v1 use both 400 and 415 status codes.
 		// See more in research: https://github.com/VictoriaMetrics/VictoriaMetrics/pull/8462#issuecomment-2786918054
 	} else if statusCode == 415 || statusCode == 400 {
-		if c.canDowngradeVMProto.Swap(false) {
-			logger.Infof("received unsupported media type or bad request from remote storage at %q. Downgrading protocol from VictoriaMetrics to Prometheus remote write for all future requests. "+
-				"See https://docs.victoriametrics.com/victoriametrics/vmagent/#victoriametrics-remote-write-protocol", c.sanitizedURL)
-			c.useVMProto.Store(false)
-		}
-
 		if encoding.IsZstd(block) {
 			logger.Infof("received unsupported media type or bad request from remote storage at %q. Re-packing the block to Prometheus remote write and retrying."+
 				"See https://docs.victoriametrics.com/victoriametrics/vmagent/#victoriametrics-remote-write-protocol", c.sanitizedURL)
 
-			block = mustRepackBlockFromZstdToSnappy(block)
+			// Block could be corrupted in some cases, e.g., due to an ungraceful shutdown of vmagent.
+			// In this case an empty slice is returned.
+			// See https://github.com/VictoriaMetrics/VictoriaMetrics/issues/9417
+			block = tryRepackBlockFromZstdToSnappy(block)
+			if len(block) != 0 {
+				if c.canDowngradeVMProto.Swap(false) {
+					logger.Infof("received unsupported media type or bad request from remote storage at %q. Downgrading protocol from VictoriaMetrics to Prometheus remote write for all future requests. "+
+						"See https://docs.victoriametrics.com/victoriametrics/vmagent/#victoriametrics-remote-write-protocol", c.sanitizedURL)
+					c.useVMProto.Store(false)
+				}
```

Comment:
I'd suggest to use some direct signal if function call was successful instead of relying on block len.
Also, this comment should be a part of function description so that it will be easier to find this if this function will be reused any time in future.

Reply:
fixed